errorcodes/errorcodes.go

```diff
@@ -0,0 +1,241 @@
+		err:            err,
+		description:    err.Error(),
+		reason:         NoReason,
+		httpCode:       503,
+		grpcStatusCode: codes.Internal,
+	}
+
+	// Now apply any requested options
+	// to overr
+	for _, option := range options {
+		option(de)
+	}
+
+	return de
+}
+
+// Error implements the error interface.
+func (c *DaprError) Error() string {
+	if c.err != nil {
+		return c.err.Error()
+	}
+	return ""
+}
+
+// Unwrap implements the error unwrapping interface.
+func (c *DaprError) Unwrap() error {
+	return c.err
+}
+
+// Description returns the description of the error.
+func (c *DaprError) Description() string {
+	if c.description != "" {
+		return c.description
+	}
+	return c.err.Error()
+}
+
+func (c *DaprError) SetDescription(description string) {
+	c.description = description
+}
+
+func (c *DaprError) SetResourceInfo(resourceInfo *ResourceInfo) {
+	c.resourceInfo = resourceInfo
+}
+
+func (c *DaprError) SetMetadata(md map[string]string) {
+	c.metadata = md
+}
+
+func WithResourceInfo(resourceInfo *ResourceInfo) ErrorOption {
+	f := func(de *DaprError) {
+		de.resourceInfo = resourceInfo
+	}
+	return f
+}
+
+func WithDescription(description string) ErrorOption {
+	f := func(de *DaprError) {
+		de.description = description
+	}
+	return f
+}
+
+func WithMetadata(md map[string]string) ErrorOption {
+	f := func(de *DaprError) {
+		de.metadata = md
+	}
+	return f
+}
+
+// *** GRPC Methods ***
+
+// FromDaprErrorToGRPC transforms the supplied error into
+// a GRPC Status. It assumes if the supplied error
+// is of type DaprError.
+// Otherwise, returns the original error.
+func FromDaprErrorToGRPC(err error) (*status.Status, error) {
+	de := &DaprError{}
+	if errors.As(err, &de) {
+		if st, ese := de.newStatusError(); ese == nil {
+			return st, nil
+		}
+		return nil, err
+	}
+
+	return nil, err
+}
+
+func (c *DaprError) newStatusError() (*status.Status, error) {
+	messages := []protoiface.MessageV1{
+		newErrorInfo(c.reason, c.metadata),
+	}
+
+	if c.resourceInfo != nil {
+		messages = append(messages, newResourceInfo(c.resourceInfo, c.err))
+	}
+
+	ste, stErr := status.New(c.grpcStatusCode, c.description).WithDetails(messages...)
+	if stErr != nil {
+		return nil, stErr
+	}
+
+	return ste, nil
+}
+
+func newErrorInfo(reason string, md map[string]string) *errdetails.ErrorInfo {
+	ei := errdetails.ErrorInfo{
+		Domain:   Domain,
+		Reason:   reason,
+		Metadata: md,
+	}
+
+	return &ei
+}
+
+func newResourceInfo(rid *ResourceInfo, err error) *errdetails.ResourceInfo {
+	return &errdetails.ResourceInfo{
+		ResourceType: rid.ResourceType,
+		ResourceName: rid.ResourceName,
+		Owner:        Owner,
+		Description:  err.Error(),
+	}
+}
+
+// *** HTTP Methods ***
+
+// FromDaprErrorToHTTP transforms the supplied error into
+// a GRPC Status and then Marshals it to JSON.
+// It assumes if the supplied error is of type DaprError.
+// Otherwise, returns the original error.
+func FromDaprErrorToHTTP(err error) (int, []byte, error) {
```

Comment:
Let's work with the runtime to make it so DaprError can be returned direclty by one of the errorResponders method in pkg/http

We already have half of the work done with the errorResponseValue interface: https://github.dev/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/pkg/http/errors.go#L20-L23

This way we can pass DaprError objects to  `respondWithError` and it will just work: https://github.dev/dapr/dapr/blob/ad5618711830510617e1b8fb2d0c6758f14ff6b9/pkg/http/responses.go#L134-L135